### PR TITLE
Extracted models + refactored to use Flask factory design pattern

### LINKED ISSUE #89: Move to the factory design pattern for easier API testing
Follow the documentation (https://flask.palletsprojects.com/en/2.0.x/patterns/appfactories/) to implement the factory pattern.

### PR DESCRIPTION
**Pull request by rudy-patel**
### What is it?
This PR extracts the models out of the main file. I also refactored the code to use the Flask factory design pattern, which allows for easier testing of the APIs (we can create separate production + testing versions of the app). A few API tests were also added to extend in the future.

Closes #89.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -1,104 +1,89 @@
...
-
-class Camera(db.Model):
-    id = db.Column(db.Integer, primary_key=True)
-    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
-    name = db.Column(db.String(50), nullable=False)
-    status = db.Column(db.Enum(CameraStatus), nullable=False)
-    mode = db.Column(db.Enum(CameraMode), nullable=False)
-    
+    loginManager.init_app(app)
+    loginManager.login_view = 'myapp.login'
```
Comment (rudy-patel) [inline main.py:27]: This is the factory pattern. The creation of the app is hidden within its own function so we can create different versions of the app. One for production, one for testing.

File: main.py [hunk 2]
```diff
@@ -1,104 +1,89 @@
...
 
+    app.register_blueprint(bp)
 
+    from models.models import db
+    db.init_app(app)
```
Comment (rudy-patel) [inline main.py:32]: The models and db has been extracted to the models package.

File: main.py [hunk 3]
```diff
@@ -1,104 +1,89 @@
...
-    id = db.Column(db.Integer, primary_key=True)
-    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
-    name = db.Column(db.String(50), nullable=False)
-    status = db.Column(db.Enum(CameraStatus), nullable=False)
-    mode = db.Column(db.Enum(CameraMode), nullable=False)
-    
+    loginManager.init_app(app)
+    loginManager.login_view = 'myapp.login'
 
+    app.register_blueprint(bp)
```
Comment (rudy-patel) [inline main.py:29]: Along with the factory pattern, we have to use **blueprints** to add/edit routes.

File: main.py [hunk 4]
```diff
@@ -1,104 +1,89 @@
...
 
-@app.route("/addEvent")
+@bp.route("/test", methods=["GET", "POST"])
+def test():
+    getData = 'get_request'
+    postData = 'post_request'
+    if(request.method == "POST"):
+        return jsonify({'data': postData})
+    elif (request.method == "GET"):
+        return jsonify({'data': getData})
```
Comment (rudy-patel) [inline main.py:57]: This is just a test route and is tested through some example tests as part of this PR. These are just placeholder, to show HOW to test JSON responses.

File: main.py [hunk 5]
```diff
@@ -1,104 +1,89 @@
...
 def testAddEvent():
     
     newEvent = Event(user_id=current_user.id, camera_id=1, type=EventType.IMAGE_CAPTURE_SUCCESS, timestamp=datetime.now())
 
-    db.session.add(newEvent)
-    db.session.commit()
+    newEvent.create()
 
-    return redirect(url_for('home'))
+    return redirect(url_for('myapp.home'))
```
Comment (rudy-patel) [inline main.py:67]: Now we need to prepend the URLs with `myapp.` to follow the factory pattern.

File: models/models.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(50), nullable=False)
+    email = db.Column(db.String(50), unique=True, nullable=False)
+    password = db.Column(db.String(80), nullable=False)
+    events = db.relationship('Event', backref='users', lazy=True)
+    cameras = db.relationship('Camera', backref='users', lazy=True)
+
+    def create(self):
+        db.session.add(self)
+        db.session.commit()
```
Comment (rudy-patel) [inline models/models.py:18]: Extracted all the DB models to their own package. The `create` method is used to communicate with the DB and insert records.

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    recorded = []
+
+    def record(sender, template, context, **extra):
+        recorded.append((template, context))
+
+    template_rendered.connect(record, app)
+    try:
+        yield recorded
+    finally:
+        template_rendered.disconnect(record, app)
```
Comment (rudy-patel) [inline tests/conftest.py:54]: These are test fixtures. These run before the tests and provide important setup steps that are often reused.

File: tests/test_routes.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+def test_test_post(client):
+    response = client.post("/test")
+    data = json.loads(response.get_data(as_text=True))
+    assert data['data'] == 'post_request'
+
+def test_test_get(client):
+    response = client.get("/test")
+    data = json.loads(response.get_data(as_text=True))
+    assert data['data'] == 'get_request'
```
Comment (rudy-patel) [inline tests/test_routes.py:16]: This is how to test for JSON responses from the Flask API.

**Review by kaden-dreger — APPROVED** [review]
Looks good

**Pull request merged**